src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -554,32 +565,74 @@ private Task ProcessHeadersFrameAsync<TContext>(IHttpApplication<TContext> appli
                 }
 
                 // Start a new stream
-                _currentHeadersStream = new Http2Stream<TContext>(application, new Http2StreamContext
-                {
-                    ConnectionId = ConnectionId,
-                    StreamId = _incomingFrame.StreamId,
-                    ServiceContext = _context.ServiceContext,
-                    ConnectionFeatures = _context.ConnectionFeatures,
-                    MemoryPool = _context.MemoryPool,
-                    LocalEndPoint = _context.LocalEndPoint,
-                    RemoteEndPoint = _context.RemoteEndPoint,
-                    StreamLifetimeHandler = this,
-                    ClientPeerSettings = _clientSettings,
-                    ServerPeerSettings = _serverSettings,
-                    FrameWriter = _frameWriter,
-                    ConnectionInputFlowControl = _inputFlowControl,
-                    ConnectionOutputFlowControl = _outputFlowControl,
-                    TimeoutControl = TimeoutControl,
-                });
-
-                _currentHeadersStream.Reset();
+                _currentHeadersStream = CreateStream(application);
+
                 _headerFlags = _incomingFrame.HeadersFlags;
 
                 var headersPayload = payload.Slice(0, _incomingFrame.HeadersPayloadLength); // Minus padding
                 return DecodeHeadersAsync(_incomingFrame.HeadersEndHeaders, headersPayload);
             }
         }
 
+        private Http2Stream CreateStream<TContext>(IHttpApplication<TContext> application)
+        {
+            Http2Stream<TContext> stream = null;
+
+            lock (_streamPool)
```

Comment:
We shouldn't need to lock in ~Create~GetStream and ReturnStream. ProcessHeadersFrameAsync is called in a single logical connection processing loop as is UpdateCompletedStreams and MakeSpaceInDrainQueue. That's how we're able to avoid any locks in Http2Connection today despite needing to accurately track the number of active streams.

Reply:
As long as Return is thread safe as well.

Reply:
ReturnStream is called from UpdateCompletedStream. I don't believe that will ever conflict with a frame being processed.